lib/reline/face.rb

```diff
@@ -0,0 +1,140 @@
+# frozen_string_literal: true
+
+class Reline::Face
+  SGR_PARAMETERS = {
+    foreground: {
+      black: 30,
+      red: 31,
+      green: 32,
+      yellow: 33,
+      blue: 34,
+      magenta: 35,
+      cyan: 36,
+      white: 37,
+      bright_black: 90,
+      gray: 90,
+      bright_red: 91,
+      bright_green: 92,
+      bright_yellow: 93,
+      bright_blue: 94,
+      bright_magenta: 95,
+      bright_cyan: 96,
+      bright_white: 97
+    },
+    background: {
+      black: 40,
+      red: 41,
+      green: 42,
+      yellow: 43,
+      blue: 44,
+      magenta: 45,
+      cyan: 46,
+      white: 47,
+      bright_black: 100,
+      gray: 100,
+      bright_red: 101,
+      bright_green: 102,
+      bright_yellow: 103,
+      bright_blue: 104,
+      bright_magenta: 105,
+      bright_cyan: 106,
+      bright_white: 107,
+    },
+    style: {
+      default: 0,
+      bold: 1,
+      faint: 2,
+      italicized: 3,
+      underlined: 4,
+      slowly_blinking: 5,
+      blinking: 5,
+      rapidly_blinking: 6,
+      negative: 7,
+      concealed: 8,
+      crossed_out: 9
+    }
+  }.freeze
+
+  class Config
+    ESSENTIAL_DEFINE_NAMES = %i(default enhanced scrollbar).freeze
+    RESET_SGR = "\e[0m".freeze
+
+    def initialize(name, &block)
+      @definition = {}
+      block.call(self)
+      ESSENTIAL_DEFINE_NAMES.each do |name|
+        @definition[name] ||= { style: :default, escape_sequence: RESET_SGR }
+      end
+    end
+
+    attr_reader :definition
+
+    def define(name, foreground: nil, background: nil, style: nil)
+      values = {}
+      values[:foreground] = foreground if foreground
+      values[:background] = background if background
+      values[:style] = style if style
+      values[:escape_sequence] = format_to_sgr(values).freeze
+      @definition[name] = values
+    end
+
+    def [](name)
+      @definition.dig(name, :escape_sequence) or raise ArgumentError, "unknown face: #{name}"
+    end
+
+    private
+
+    def sgr_rgb(key, value)
+      return nil unless rgb_expression?(value)
+      case key
+      when :foreground
+       "38;2;"
+      when :background
+        "48;2;"
+      end + value[1, 6].scan(/../).map(&:hex).join(";")
+    end
+
+    def format_to_sgr(values)
+      sgr = "\e[" + values.map do |key, value|
+        case key
+        when :foreground, :background
+          case value
+          when Symbol
+            SGR_PARAMETERS[key][value]
+          when String
+            sgr_rgb(key, value)
+          end
+        when :style
+          [ value ].flatten.map { |style_name| SGR_PARAMETERS[:style][style_name] }
```

Comment:
If style contains `:default`, foreground and background is reset.
I think this is a little confusing
```ruby
# red, cyan, bold
face.define :enhanced, background: :red, foreground: :cyan, style: :bold
# default background, default foreground
face.define :enhanced, background: :red, foreground: :cyan, style: :default
```

Other part of this pull request looks good to me :+1:

Reply:
Yap, that's why I originally named `:normal_line` instead of `:default` 🥺

Reply:
...How about `style :reset` ?

Reply:
RESET_SGR is always set at `sgr == RESET_SGR ? RESET_SGR : RESET_SGR + sgr`, so using `SGR_PARAMETERS[:style][:default]` or `SGR_PARAMETERS[:style][:reset]` here has no effect. I thought excluding `:default` is missing.
```ruby
[ value ].flatten.flter_map do |style_name|
   SGR_PARAMETERS[:style][style_name] if style_name != :default
end
```

Reply:
Okay, I got what you're saying.
Fortunately, I guess changing `style: :default` to `style: :reset` made the situation better.
Parameters before `:reset` will be reset while parameters after `:reset` are not ignored because the order of the escape sequences will be the same as the order of the parameters in the `define` method.
example:
```
format_to_sgr({foreground: :white, style: :reset, background: :red})
=> "\e[0m\e[37;0;41m"
```

Since Reline users are programmers, I think there is no problem with them writing freely and taking responsibility for it.

Reply:
Looks good.
Perhaps `face.define :default, style: :default` in `reline.rb` should be `face.define :default, style: :reset`?
It seems that `invalid SGR parameter` check is skipped for style

Reply:
You are right 👍  https://github.com/ruby/reline/pull/552/commits/67a8497ba1e29dc348c88667f589f0fcb2a79d48

And I was wrong one more thing 😓https://github.com/ruby/reline/pull/552/commits/c02205c0c1a17d7ca17f368b9e6dc635976a1235